utils/writeable/src/impls.rs

```diff
@@ -8,84 +8,82 @@ use core::convert::TryFrom;
 use core::fmt;
 use core::str;
 
-impl Writeable for u8 {
-    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
-        let mut buf = [b'0'; 3];
-        let mut n = *self;
-        let mut i = 3usize;
-        while n != 0 {
-            i -= 1;
-            buf[i] = b'0' + (n % 10);
-            n /= 10;
-        }
-        if i == 3 {
-            debug_assert_eq!(*self, 0);
-            i = 2;
-        }
-        let s = unsafe { str::from_utf8_unchecked(&buf[i..]) };
-        sink.write_str(s)
-    }
+macro_rules! impl_write_num {
+    ($u:ty, $i:ty) => {
+        impl Writeable for $u {
+            fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+                // A number x requires ⌊log₁₀(x)⌋ + 1 digits in base 10, so we need a
+                // buffer of length ⌊log₁₀($u::MAX)⌋ + 1. As logarithms aren't const,
+                // we have to simplify a bit:
+                // ⌊log₁₀($u::MAX)⌋ = ⌊log₁₀(2ᵇ - 1)⌋ ≛ ⌊log₁₀(2ᵇ)⌋ = ⌊b log₁₀(2)⌋
+                // (*) holds because there is no integer in [log₁₀(2ᵇ - 1), log₁₀(2ᵇ)],
+                // if there were, there'd be some 10ⁿ in [2ᵇ - 1, 2ᵇ], but it can't be 
+                // 2ᵇ - 1 due to parity nor 2ᵇ due to prime factors.
+                let mut buf = [b'0'; (<$u>::BITS as f64 * core::f64::consts::LOG10_2) as usize + 1];
+                let mut n = *self;
+                let mut i = buf.len();
+                while n != 0 {
+                    i -= 1;
+                    buf[i] = b'0' + u8::try_from(n % 10).expect("<10");
+                    n /= 10;
+                }
+                if i == buf.len() {
+                    debug_assert_eq!(*self, 0);
+                    i -= 1;
+                }
+                let s = unsafe { str::from_utf8_unchecked(&buf[i..]) };
+                sink.write_str(s)
+            }
 
-    fn write_len(&self) -> LengthHint {
-        if *self < 10 {
-            LengthHint::exact(1)
-        } else if *self < 100 {
-            LengthHint::exact(2)
-        } else {
-            LengthHint::exact(3)
+            fn write_len(&self) -> LengthHint {
```

Comment:
Thought: I am aware of at least 3 ways to compute this.

1. Divide by 10 in a loop, as you are doing
2. Lookup table, similar to what we had before (this can be a binary search to speed it up)
3. Computing `self.log10() + 1`

I'd be interested to see both performance and code size impacts of those three approaches.

If you don't want to do this now, please file an issue for someone to take in the future.

Reply:
I don't think we should use `.log10()`. If there's no hardware floating point unit it will probably run some Newton approximation in ints to an accuracy we don't need which will get pretty expensive.

We can take a cheap floor-log2 by inspection the highest bit, then we're off by at most one digit, which we can fix with one `pow` invocation.

Reply:
In my rough benchmarking (0..25_000_000) I see
1. Divide by 10 in a loop: 1.9s
2. Binary search over hardcoded array: 3.8s (+ probably biggest code size)
3. `log10`: 1.8s on `x86_64`, but also `no_std` doesn't seem to include logarithms
4. `log10` approximation: 1.2s